### DISCUSSION TITLE
add hook to do something when All DAGs before each run

### DISCUSSION THREAD

**Discussion by chuxiangfeng** [apache/airflow#27712] [General] [answered]
### Description

like for each dag add on_success_callback or on_failure_callback, i want one hook to do some thing before all DAGs each run

### Use case/motivation

Add a function hook can be called before all DAGs each run ， in function can get some infos about dag， like dag_id，task infos for dag...

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by potiuk** [comment, marked as answer]
You have https://airflow.apache.org/docs/apache-airflow/stable/concepts/cluster-policies.html - check if it is ok for you to do stuff (maybe you will have to do some state management extrnally if you want to limit the number of times the cluster policy is triggered).

There is also a work in progress of setup/tearDown mechanism for DAGs that will be pubished in devlist in the coming weeks and AIP proposal https://lists.apache.org/list.html?dev@airflow.apache.org  - subscribe to the devlist and watch the devlist and see if you can comment on the proposal that will be published there, it will likely address your expectations - but better approach will be that you read the proposal that will come and see if this is somethign that fits your needs./ .

**Reply by chuxiangfeng** [reply]
think you, i will try cluster-policies later.

**Reply by chuxiangfeng** [reply]
> You have https://airflow.apache.org/docs/apache-airflow/stable/concepts/cluster-policies.html - check if it is ok for you to do stuff (maybe you will have to do some state management extrnally if you want to limit the number of times the cluster policy is triggered).
> 
> There is also a work in progress of setup/tearDown mechanism for DAGs that will be pubished in devlist in the coming weeks and AIP proposal https://lists.apache.org/list.html?dev@airflow.apache.org - subscribe to the devlist and watch the devlist and see if you can comment on the proposal that will be published there, it will likely address your expectations - but better approach will be that you read the proposal that will come and see if this is somethign that fits your needs./ .

It does what I want it to do, but unfortunately it doesn't fire every time the dag runs or the dag structure changes, and seems to be executed only on the first launch

**Reply by potiuk** [reply]
Then If you think you need it, you will want to start a discussion in the devlist and likly prepare an AIP (airflow improvement proposal) - and likely either i.plement it or find someone who will lead it to completion - you will find links at community page of our  website. Having more hooks like you think is a new thing thatt might have serious performance implications and has to be very well though and discussed -- this is a whole new functionality that has never been raised before and you will need to justify and convince the others, propose and lead the AIP to make something like to happen - unless form the discussion you will need someone who would like to do it.

Just to let you know how the workflow looks like in Apache projects as you might be not aware that such. IG functionity required such approach. Features / GitHub issues are Good for small things that do not introduce such serious internal changes.

**Reply by potiuk** [reply]
Setup an teardown are likely only address task execution steps..

**Reply by potiuk** [reply]
Also another comment - if you are about structure changes, that has to be also part of DAG versioning AIP which has been discussed but the discussions never completed and it is not even approved - but it is prerequisite IMHO to notify about structural DAG changes https://cwiki.apache.org/confluence/plugins/servlet/mobile?contentId=158868919#content/view/158868919
